src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java

```diff
@@ -91,44 +96,54 @@ public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst)
     @Override
     public int compareOrder(NodeInfo other) {
         int result = 0;
-        if (other instanceof AbstractNode) {
-            result = getDepth() - ((AbstractNode) other).getDepth();
+        if (other instanceof ElementNode) {
+            result = Integer.compare(depth, ((ElementNode) other).depth);
             if (result == 0) {
-                final ElementNode[] children = getCommonAncestorChildren(other);
-                result = children[0].indexAmongSiblings - children[1].indexAmongSiblings;
+                result = compareCommonAncestorChildrenOrder(this, other);
```

Comment:
is this update required to fix stackoverflow ? if this is not related better to move to separate commit of the same PR.

Reply:
done. I'll push "optimizations" commit once the CI passed.